File: Compiler/src/compiler/Variable.java
package compiler;

import util.CompilationResult;
import util.DataType;
import util.JenaUtil;

import java.util.ArrayList;
import java.util.List;

/**
 * Переменная, вводимая некоторыми операторами
 */
public class Variable implements Operator {

    /**
     * Имя переменной
     */
    private final String name;

    /**
     * Тип данных переменной
     */
    private final DataType dataType;

    /**
     * Конструктор
     * @param name Имя переменной
     * @param dataType Тип данных переменной
     */
    public Variable(String name, DataType dataType) {
        this.name = name;
        this.dataType = dataType;
    }

    @Override
    public List<List<DataType>> argsDataTypes() {
        return new ArrayList<>();
    }

    @Override
    public DataType resultDataType() {
        return dataType;
    }

    @Override
    public CompilationResult compile() {
        return new CompilationResult(JenaUtil.genRuleVar(name), "", "");
    }
}

File: Compiler/src/compiler/Variable.kt
package compiler

import compiler.util.CompilationResult
import util.DataType
import util.JenaUtil

/**
 * Переменная, вводимая некоторыми операторами
 * @param name Имя переменной
 */
class Variable(
    private val name: String
) : Operator {

    override val args: List<Operator>
        get() = emptyList()

    override val argsDataTypes: List<List<DataType>>
        get() = emptyList()

    override val resultDataType: DataType
        get() = DataType.Object

    override fun compile(): CompilationResult =
        CompilationResult(value = JenaUtil.genVar(name))

    override fun clone(): Operator = Variable(name)

    override fun clone(newArgs: List<Operator>): Operator {
        require(newArgs.isEmpty()) { "Для переменной аргументы не требуются." }
        return clone()
    }
}